Fix front-end dependencies handling

WouterJ opened the pull request:
Closes #81 and #85

There are quite some changes in here (and also a pretty bold move, which I hope you agree with). Let sum them up:

* The `CmfTreeBrowserBundle:Base:tree.html.twig` template is moved to a `CmfTreeBrowserBundle:Base:scripts.html.twig` file, which is included in the `...:Base:tree.html.twig` template. By default, it'll use the front-end packages that are in the bundle. If someone is using bower for their symfony application, they install the `symfony-cmf-tree-browser-bundle` bower package and override the `...:Base:tree.html.twig` file, like so:

  ```
  {% include 'CmfTreeBrowserBundle:Base:scripts.html.twig' with {bowerPackagePath: 'bower_components'} %}
  ```
  This will then pick the files it needs from `web/bower_components` instead of the bundle.

* Instead of including all files installed by Bower in the bundle, only the files that are used by the bundle are committed.

* The GulpJS task runner is used to avoid Assetic usage and make the above change more easy. This means that before every release, one has to run `gulp`. This will then: install the deps, pick the files it needs and moves them to `Resources/public/vendor`, minify all JS files in the bundle and concat them into one file: `cmf_tree_browser.min.js`.

dbu commented on Resources/views/Base/tree.html.twig:
```diff
@@ -1,11 +1 @@
-
-<link rel="stylesheet" href="{{ asset('bundles/cmftreebrowser/vendor/fancytree/dist/skin-win8/ui.fancytree.min.css') }}" media="all"/>
-<link rel="stylesheet" href="{{ asset('bundles/cmftreebrowser/css/fontawesome-style.css') }}" media="all"/>
+{% include 'CmfTreeBrowserBundle:Base:scripts.html.twig' with {bowerPackagePath: 'bundles/cmftreebrowser/vendor'} %}
```
why call this path bower and not just assetsBasePath?

WouterJ replied:
yeah, we can go with that.

dbu commented:
so the default fallback is that all dependencies are git checked into this repository?

and when using bower, those files are not used but bower fetches the right thing? how is our own bower file and gulp task included in a symfony application? do we provide this bundle as a bower package? (sorry, i am not very familiar with bower/gulp yet so i am quite confused :-)

could we instead of committing jquery and jquery-ui just use https://code.jquery.com/jquery-2.1.3.min.js and https://code.jquery.com/ui/1.11.4/jquery-ui.min.js instead? that would seem more sane to me than copying jquery into this repository. i guess for fancytree, there is no such reliable CDN available that we could use?

btw, the tree is embedded into a page, right? what if that page already has jquery loaded? do we need to make loading jquery optional, or can we just document that the base template using the tree must handle jquery inclusion?

WouterJ commented:
> so the default fallback is that all dependencies are git checked into this repository?

Yes (but only the files we require, so no README's and such)

> and when using bower, those files are not used but bower fetches the right thing?

Indeed, then one makes bower install the required packages in, for instance, `/web/assets/vendor`. In this case, you always want to use that file, as there is a high change it's more up to date than the one included in each release.

> how is our own bower file and gulp task included in a symfony application?

Gulp tasks are just to make our life as Bundle devs easier, the user of the bundle doesn't have to worry about it. Instead of running `bower install` and removing all unrelated installed files, we now just have to run `gulp pack` and everything will be ready for a release.

Our own bower file is like `composer.json`: It tells everyone depending on it which packages it needs and which versions it supports.

> do we provide this bundle as a bower package?

Yes: `symfony-cmf-tree-browser-bundle`. It uses this repository. Bower just fetches the complete repository and then removes all files in the `ignore` of `bower.json`. In our case, this means that all files will be removed. This package is purely for providing the dependencies.

> could we instead of committing jquery and jquery-ui just use https://code.jquery.com/jquery-2.1.3.min.js and https://code.jquery.com/ui/1.11.4/jquery-ui.min.js instead?

I'm -1 for depending on a CDN: It doesn't work when you're offline. Besides that, using bower files, the user can chose which version to use and we'll make sure we have correct version constraints.

> what if that page already has jquery loaded?

That's exactly what these 2 lines do in `CmfTreeBrowserBundle:Base:scripts.html.twig`:

```html
<script>window.jQuery || document.write(unescape('%3Cscript src="{{ asset(bowerPackagePath ~ '/jquery/dist/jquery.min.js') }}"%3E%3C/script%3E'))</script>
<script>jQuery.ui || document.write(unescape('%3Cscript src="{{ asset(bowerPackagePath ~ '/jquery-ui/jquery-ui.min.js') }}"%3E%3C/script%3E'))</script>
```
It says, check if `window.jQuery` is defined (`jQuery`), if it isn't `document.write` a script tag importing the jQuery script from the defined path. Same for `jQuery.ui`. (crazy escape sequences are required as every `</script>` in inline JS, even when it's in a string, is assumed to close the current script element).

dbu commented:
thanks a lot wouter, see it more clearly now.

what i am still unclear: why do we commit `.min` files of the js we provide in this repository? in my understanding, committing "compiled" files is a big problem as you need to make sure they actually are correctly recompiled on every commit. can't we just use the non-min version for development, and have a build that provides minified releases for bower?

and why do we not provide the javascript provided by this bundle as bower as well? if people use a bower-gulp workflow, i think that would be more convenient for them.

WouterJ commented:
> what i am still unclear: why do we commit .min files of the js we provide in this repository? in my understanding, committing "compiled" files is a big problem as you need to make sure they actually are correctly recompiled on every commit. can't we just use the non-min version for development, and have a build that provides minified releases for bower?

If we do not provide a min file in development, you would need to include other files when using a dev version of the TreeBrowser or a  stable release. I'm -1 on that. For each stable release, the min file is updated and for dev versions, we have to remember to update it regularly imo

> and why do we not provide the javascript provided by this bundle as bower as well?

I see no case in which you need the javascript files, but not the bundle.

WouterJ commented:
Updated this PR:

* Renamed option to set the basepath in `scripts.html.twig`
* Added adapter option
* Removed minified file, we'll have to discuss how to do this in another issue

I propose to merge this PR as is, it would allow #88 to be merged.

dbu commented:
as seems usual these days, this fails with the missing --force option. i would love to see the tests working before merging.

there is a PR to backport the fix with dbal to 2.4 which hopefully gets merged tomorrow. then we can relax the version requirement for jackalope-doctrine-dbal again and the right version should be installed.

any chance you could adjust the Testing component to handle the missing --force parameter?

WouterJ commented:
@dbu I tried fixing it: https://github.com/symfony-cmf/Testing/pull/98 (I'm not really a bash expert though)

WouterJ commented:
build is fixed now
